### DISCUSSION TITLE
strongSwan 5.9.8, no capable fetcher found did not receive a valid scep response: HTTP 0

### DISCUSSION THREAD

**Discussion by thrdgeek** [strongswan/strongswan#2512] [Q&A]
ipsec statusall
Status of IKE charon daemon (strongSwan 5.9.8, Linux 6.8.0-40-generic, x86_64):
  uptime: 4 days, since Oct 23 08:59:16 2024
  malloc: sbrk 2969600, mmap 0, used 830768, free 2138832
  worker threads: 12 of 16 idle, 4/0/0/0 working, job queue: 0/0/0/0, scheduled: 0
  loaded plugins: charon-systemd addrblock aes aesni agent certexpire connmark eap-aka eap-dynamic eap-gtc eap-identity eap-radius error-notify farp forecast led lookip tnc-tnccs unity xauth-eap xauth-pam des rc2 sha2 sha1 md5 mgf1 random nonce x509 revocation constraints pubkey pkcs1 pkcs7 pkcs12 pgp dnskey sshkey pem openssl pkcs8 fips-prf gmp curve25519 xcbc cmac hmac kdf gcm drbg curl attr kernel-netlink resolve socket-default stroke vici updown xauth-generic counters
Listening IP addresses:
  x.x.x.x
Connections:
Security Associations (0 up, 0 connecting):
  none

Curl is presently loaded but pki is unable to fetch the certificate.  What else could be going on?

**Comment by tobiasbrunner** [comment]
> Curl is presently loaded but pki is unable to fetch the certificate. What else could be going on?

Is it loaded in pki (see `pki --help`)?

**Reply by thrdgeek** [reply]
yes
pki --help
strongSwan 5.9.8 PKI tool
loaded plugins: aes des rc2 sha2 sha1 md5 mgf1 random x509 revocation pubkey pkcs1 pkcs7 pkcs12 dnskey sshkey pem openssl pkcs8 gmp curve25519 hmac kdf drbg curl

**Reply by thrdgeek** [reply]
Here is the string I am submitting.

pki --scep --url http://scepserver.internal.com/ejbca/publicweb/apply/scep/Ubuntu/pkiclient.exe --password 12345678 --in /home/user/.cisco/certificates/client/private/client-key.pem --dn CN=user,OU=group,OU=country,DC=site,DC=corp,DC=company,DC=com --san user@site.corp.company.com --cacert-enc /etc/ipsec.d/cacerts/caCert-ra-1-1.der --cacert-sig /etc/ipsec.d/cacerts/caCert-ra-1-1.der --outform pem --debug 4

**Reply by tobiasbrunner** [reply]
Please post the complete output of that command (maybe reduce the debug level to 2). And please also post the output of `ipsec listplugins | grep -1 FETCHER`.

**Reply by thrdgeek** [reply]
sudo ipsec listplugins | grep -1 FETCHER
[sudo] password for user: 
        CERT_DECODE:X509 (soft)
        FETCHER:(null) (soft)
constraints:
--
curl:
    FETCHER:file://
    FETCHER:ftp://
    FETCHER:http://
    FETCHER:https://
        CUSTOM:openssl-threading

**Reply by tobiasbrunner** [reply]
That looks OK. What about the log output of the other command?

**Reply by thrdgeek** [reply]
builder L0 CRED_PRIVATE_KEY - RSA of plugin 'pkcs1'
builder L0 CRED_PRIVATE_KEY - RSA of plugin 'openssl'
builder L0 CRED_PRIVATE_KEY - RSA of plugin 'pkcs8'
builder L0 CRED_PRIVATE_KEY - RSA of plugin 'gmp'
builder L0 CRED_PRIVATE_KEY - RSA of plugin 'pem'
  file content is not binary ASN.1
  -----BEGIN RSA PRIVATE KEY-----
  -----END RSA PRIVATE KEY-----
builder L1 CRED_PRIVATE_KEY - RSA of plugin 'pkcs1'
L0 - RSAPrivateKey:
L1 - version:
L1 - modulus:
L1 - publicExponent:
L1 - privateExponent:
L1 - prime1:
L1 - prime2:
L1 - exponent1:
L1 - exponent2:
L1 - coefficient:
builder L2 CRED_PRIVATE_KEY - RSA of plugin 'pkcs1'
builder L2 CRED_PRIVATE_KEY - RSA of plugin 'openssl'
builder L0 CRED_PUBLIC_KEY - RSA of plugin 'pkcs1'
L0 - RSAPublicKey:
L1 - modulus:
L1 - publicExponent:
builder L1 CRED_PUBLIC_KEY - RSA of plugin 'pkcs1'
builder L1 CRED_PUBLIC_KEY - RSA of plugin 'dnskey'
builder L1 CRED_PUBLIC_KEY - RSA of plugin 'openssl'
sending scep request to 'http://scepserver.com/ejbca/publicweb/apply/scep/Ubuntu/pkiclient.exe'
unable to fetch from http://scepserver.com/ejbca/publicweb/apply/scep/Ubuntu/pkiclient.exe?operation=GetCACaps, no capable fetcher found
did not receive a valid scep response: HTTP 0

**Reply by tobiasbrunner** [reply]
Hm, this looks fine and if the `curl` plugin is loaded in `pki` with the same features reported by the daemon, then it's weird this fails. Have you configured any options in strongswan.conf or files it includes? Did you build strongSwan yourself? 


Also be aware that the version you are using doesn't correctly handle those CA capabilities (see #1607, which was fixed with 5.9.11).

**Reply by thrdgeek** [reply]
I have not added anything to strongswan.conf:
```
cat /etc/strongswan.conf
# strongswan.conf - strongSwan configuration file
#
# Refer to the strongswan.conf(5) manpage for details
#
# Configuration changes should be made in the included files

charon {
	load_modular = yes
	plugins {
		include strongswan.d/charon/*.conf
	}
}

include strongswan.d/*.conf
```
I did build this myself.
`./configure   --prefix=/usr   --sysconfdir=/etc   --enable-openssl   --enable-pki   --enable-systemd   --enable-curl`

**Reply by tobiasbrunner** [reply]
> I did build this myself.

Then why did you build such an old version? Also, if those are your complete configure options, you apparently have some binary packages installed (see the list of plugins by the daemon, which lists a lot of plugins that are disabled by default).  That's never a good idea and could cause all sorts of weird issues. Always make sure to completely remove all strongSwan-related binary packages before installing a custom build.

**Comment by thrdgeek** [comment]
Not sure.  I am trying it with 5.9.14 now.

**Reply by thrdgeek** [reply]
I have success with the latest version.  Well at least I can communicate with the scep server now.  Thank you.
